README.md
# MMM-LesJoiesDuCode

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

MMM-LesJoiesDuCode provides the last posts from [Les Joies Du Code](https://lesjoiesducode.fr), or [TheCodingLove](https://thecodinglove.com) (The english version).


![Demo](https://github.com/Tomadelostacos/MMM-LesJoiesDuCode/blob/master/images/demo.gif)
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2FTomadelostacos%2FMMM-LesJoiesDuCode.svg?type=shield)](https://app.fossa.io/projects/git%2Bgithub.com%2FTomadelostacos%2FMMM-LesJoiesDuCode?ref=badge_shield)

## Using the module

To install, clone this repo into `~/MagicMirror/modules` directory. Then move in the folder and install required libraries
```
cd ~/MagicMirror/modules
git clone https://github.com/Tomadelostacos/MMM-LesJoiesDuCode
cd MMM-LesJoiesDuCode
npm install
```

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
  modules: [
    {
      module: "MMM-LesJoiesDuCode",
      position: "top_center",
      config: {
        updateInterval: 3600000,
        rotateInterval: 60000,
        grayscale: true,
        language: "fr"
      }
    }
  ]
}
```

## Configuration options


| Option           | Description
|----------------- |-----------
| `updateInterval` | *Optional* How frequently you want it to update (in ms). By default: One hour
| `rotateInterval` | *Optional* Rotate time between each GIFs (in ms). By default: 60 secs
| `grayscale`      | *Optional* Makes all GIFs black and white (Looks better i think). By default: true
| `language`       | *Optional* "fr" for French, and "en" for English By default: "fr"



## License
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2FTomadelostacos%2FMMM-LesJoiesDuCode.svg?type=large)](https://app.fossa.io/projects/git%2Bgithub.com%2FTomadelostacos%2FMMM-LesJoiesDuCode?ref=badge_large)


CHANGELOG.md
# MMM-LesJoiesDuCode Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.1.0] - 05/2023 - @bugsounet: Global review of the code
 * replace request by axios
 * use a seed request for don't use cache when download result
 * force playing video on display new GIF
 * move fetch interval to node_helper
 * fetch all last posts (generaly 10)
 * add rotateInterval for ramdom display

## [1.0.0] - Unreleased

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-LesJoiesDuCode
 *
 * By AceLan Kao
 * Review by @bugsounet (05/2023)
 * MIT Licensed
 */

var NodeHelper = require('node_helper')
const axios = require("axios")

module.exports = NodeHelper.create({
  start: function () {
    this.data_url = "https://lesjoiesducode.fr/wp-json/wp/v2/posts"
  },

  getData: function () {
    let results = []
    axios({ url: this.data_url+"?seed="+Date.now() })
      .then(response => {
        if (response.data.length) {
          response.data.forEach(ljdc => {
            let content = {}
            content.title = ljdc.title.rendered
            content.content = ljdc.content.rendered
            results.push(content)
            console.log("[LJDC] Fetch:", ljdc.title.rendered)
          })
          this.sendSocketNotification("DATA_RESULTS", results)
        }
      })
      .catch(error => {
        console.error("[LJDC] Fetch data error:", error)
      })
  },

  socketNotificationReceived: function(notification, payload) {
    if (notification == "INIT") this.initialize(payload)
  },

  initialize: function(payload) {
    console.log('[LJDC] MMM-LesJoiesDuCode started')
    this.config= payload
    if (this.config.language == "en") this.data_url = "https://thecodinglove.com/wp-json/wp/v2/posts"
    this.getData()
    setInterval(() => { this.getData() }, this.config.updateInterval)
  }
});


MMM-LesJoiesDuCode.js
/* global Module */

/* Magic Mirror
 * Module: MMM-LesJoiesDuCode
 *
 * By Thomas Favre (TomaDeLosTacos)
 * Review by @bugsounet (05/2023)
 * MIT Licensed.
 */

"use strict";

Module.register("MMM-LesJoiesDuCode", {
  defaults: {
    updateInterval: 3600000,
    rotateInterval: 60000,
    grayscale: false,
    language: "fr"
  },

  start: function() {
    this.results= [ { title: "loading...", content: ""} ]
    this.rotate = null
    this.sendSocketNotification("INIT", this.config)
  },

  getDom: function() {
    var wrapper = document.createElement("div")
    wrapper.id = "ljdc-wrapper"

    var ljdcWrapper = document.createElement("div")
    ljdcWrapper.id = "ljdc-info"

    var flexWrapper = document.createElement("div")
    flexWrapper.id = "flex-wrapper"

    var ljdcName = document.createElement("p")
    ljdcName.innerHTML = this.results[0].title
    ljdcName.id = "ljdc-name"
    wrapper.appendChild(ljdcName)

    var ljdcContent = document.createElement("div")
    ljdcContent.innerHTML = this.results[0].content

    ljdcContent.id = "ljdc-content"
    if(this.config.grayscale) ljdcContent.id = "ljdc-content-grayscale"

    ljdcWrapper.appendChild(ljdcContent)

    flexWrapper.appendChild(ljdcWrapper)

    wrapper.appendChild(flexWrapper)

    return wrapper
  },

  socketNotificationReceived: async function(notification, payload) {
    if (notification == "DATA_RESULTS") {
      clearInterval(this.rotate)
      this.rotate = null
      this.results = payload
      this.hide(500, () => {this.reFreshContent()}, {lockString: "ljdc"})
      this.rotate = setInterval(() => {
        this.hide(500, () => {this.reFreshContent()}, {lockString: "ljdc"})
      }, this.config.rotateInterval)
    }
  },

  reFreshContent: function () {
    let ramdom = this.getRandomInt(this.results.length)
    let name = document.getElementById("ljdc-name")
    name.innerHTML = this.results[ramdom].title
    let content = document.getElementById("ljdc-content")
    content.innerHTML = this.results[ramdom].content
    let video = content.querySelector("video")
    if (video) video.play()
    this.show(500, () => {}, {lockString: "ljdc"})
  },

  getStyles: function() {
    return [this.file('MMM-LesJoiesDuCode.css')]
  },

  getRandomInt: function(max) {
    return Math.floor(Math.random() * max)
  }
});
